shell/platform/fuchsia/flutter/gfx_external_view_embedder.cc

```diff
@@ -271,6 +274,38 @@ void GfxExternalViewEmbedder::SubmitFrame(
     }
   }
 
+  // Render the recorded SkPictures into the surfaces.
+  {
+    TRACE_EVENT0("flutter", "RasterizeSurfaces");
+
+    for (const auto& surface_index : frame_surface_indices) {
+      TRACE_EVENT0("flutter", "RasterizeSurface");
+
+      FML_CHECK(surface_index.second < frame_surfaces.size());
+      SurfaceProducerSurface* surface =
+          frame_surfaces[surface_index.second].get();
+      FML_CHECK(surface != nullptr);
+
+      sk_sp<SkSurface> sk_surface = surface->GetSkiaSurface();
+      FML_CHECK(sk_surface != nullptr);
+      FML_CHECK(SkISize::Make(sk_surface->width(), sk_surface->height()) ==
+                frame_size_);
+      SkCanvas* canvas = sk_surface->getCanvas();
+      FML_CHECK(canvas != nullptr);
+
+      const auto& layer = frame_layers_.find(surface_index.first);
+      FML_CHECK(layer != frame_layers_.end());
+      sk_sp<SkPicture> picture =
+          layer->second.recorder->finishRecordingAsPicture();
+      FML_CHECK(picture != nullptr);
+
+      canvas->setMatrix(SkMatrix::I());
```

Comment:
The complex part I was talking about before....

Lines 302-306 need to happen after the Present() below

We pipeline the Skia rendering work with the scenic IPC, and the fences passed to Present() make sure Scenic delays internally until Skia is finished

Moving the Skia rendering work up here will likely cause a performance regression


You really only need to move "finishRecordingAsPicture" up here, in order to populate the RTree.  Can you break this part into 2 pieces, before and after Present()?  It would also help to add a comment below, on the Skia code after Present, explaining the pipelining (my bad for not doing that originally)

Reply:
Thanks for the explanation. I broke up the pre- and post-Present() operations as you suggested.

I had to cache the sk_sp<SkPicture> returned by `finishRecordingAsPicture` at the beginning of the method, so that I could draw it to the canvas after the Present() call at the end of the method. Is that safe to do?

Reply:
Yeah that should be safe to do.

`finishRecordingAsPicture` doesn't actually do any GPU work, it just records the commands from the SkRecordingCanvas into an SkPicture (a type of command buffer).

The canvas->drawPicture after Present() is what actually does the GPU work